Potential idea for handling user credits in plan

ZzRanger opened the pull request:
After having many long-winded conversations about how to handle user credits in both the "Credits" and "Plans" page, I have a proposed solution.

## **Changes**
1. Change "Credits" page to "Course History"
2. Give users the <ins>option</ins> to add credits when creating their plan. Templates will never load user credits
3. Add a UI indicator for semesters where it's not "synced" with the courses from a user's course history 

## Below are gifs of the proposed changes

### Creating a new plan
![Feb-03-2023 13-13-38](https://user-images.githubusercontent.com/47403443/216687744-23f772cf-3bfa-4aaf-be6b-358f08974824.gif)

### Creating a template
![Feb-03-2023 13-13-29](https://user-images.githubusercontent.com/47403443/216687753-9a805a28-9137-4cb7-ae08-7e11f1ab579f.gif)


Let me know what you guys think! I hopping this solution straddles the line between preventing users from making bad decisions while providing them with freedom and flexibility.

ZzRanger committed:
fix degree templates

ZzRanger committed:
convert major to degreeName for validator

ZzRanger committed:
fix seed to include TemplateDataItems

ZzRanger committed:
redo how credits is handled

ZzRanger committed:
remove disabled courses & semester tiles

ZzRanger committed:
fix height on semestertile

ZzRanger committed:
add daisy-ui

ZzRanger committed:
add helper function

ZzRanger committed:
initial implementation of new credits

ZzRanger committed:
update text for Sidebar & Credits page + add tooltip

ZzRanger committed:
ssr pg_2 in onboarding

ZzRanger committed:
fix grid css

ZzRanger committed:
add space between prefix & number for courses transcript

ZzRanger committed:
show transfer credits on plan

ZzRanger committed:
add toggle for transfer credits

ZzRanger committed:
update majors list to include bachelor type & remove masters

ZzRanger committed:
briefly build new template plan

jasonappah reviewed (changes requested):
Liking the look of this so far, I think this is a good compromise based on the discussions we've had :)

Concerns:
- When moving a completed course from its original semester to another semester it loses its golden color - is this intended? I think it should keep its color when moved so that it's still clear that the course was populated from course history.
- I noticed an error when I click on the Create Plan button in the custom plan creation modal, although the plan does get created successfully.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/30608521/216702016-32abe53b-6a53-4d9e-bf14-5cfc6020a387.png">

Some other thoughts that I'd like to discuss at some point (non-blocking, could be addressed in another PR later down the road):
- Is there some way to add a legend for what the gold color means somewhere, just so it's clearer to the user that it's a transfer credit? We could try making the "Show Transfer Credits" toggle the same color as the courses.
- On the note of the colors, the blue on gold combo for completed courses is a bit hard to read imo, might be worth changing if we can...
- It'd also be nice if there was a way to see what courses are misplaced when the "Courses not synced with course history" icon gets displayed. Not sure how we'd go about that yet, would be happy to hear anyone's thoughts on these points :)

ZzRanger committed:
fix CustomPlan bug

ZzRanger committed:
show selected major on PlanCard

ZzRanger commented:
> Liking the look of this so far, I think this is a good compromise based on the discussions we've had :)
> 
> Concerns:
> 
>     * When moving a completed course from its original semester to another semester it loses its golden color - is this intended? I think it should keep its color when moved so that it's still clear that the course was populated from course history.

That's a good point. Gold is supposed to denote transfer credit and I'll keep the indicator even if you change the semester.

>     * I noticed an error when I click on the Create Plan button in the custom plan creation modal, although the plan does get created successfully.

This is because I forgot to commit and push. Fixed now :)


> Some other thoughts that I'd like to discuss at some point (non-blocking, could be addressed in another PR later down the road):
> 
>     * Is there some way to add a legend for what the gold color means somewhere, just so it's clearer to the user that it's a transfer credit? We could try making the "Show Transfer Credits" toggle the same color as the courses.

Yeah good point it's not super clear. That's a good idea to make it the same color. Kevin suggested using tags instead so I might give that a go and see how it looks. I can also add a tooltip to mention that it's a transfer credit.

>     * On the note of the colors, the blue on gold combo for completed courses is a bit hard to read imo, might be worth changing if we can...

Ye sure I just did a random color xD

>     * It'd also be nice if there was a way to see what courses are misplaced when the "Courses not synced with course history" icon gets displayed. Not sure how we'd go about that yet, would be happy to hear anyone's thoughts on these points :)

I might just have a tooltip that specifies it on the Semester level which courses are misplaced. See how that looks but I'm down to change it if we have other ideas.

ZzRanger committed:
update error validation logic

ZzRanger commented:
## Changes Made
1. Added "tags" C & T which signify Complete and Transfer Credit respectively. This replaces the golden background.
2. Replaced sync indicator on Semester card with an error count along with descriptions of error type.
3. Added specific course error messages to courses.

## How Logic Works
1. Check for whether course has been completed/is a transfer. If so, we give it the appropriate tag
2. We show an error on the semester and course level for the following cases:
 - semester **with course history** contains an extra course or is missing a course
 - a completed course is placed in another semester

We're currently assuming that if there is **a course** for a semester in Course History, that semester should **only** contain courses from course history, otherwise there is an error message that appears. Where this gets interesting is for new students coming into UTD with transfer credit (which will be listed in their first semester of their plan). 

Should we operate under the assumption that courses the user is currently taking should be listed in course history? If not, this logic may need to be reworked. Hope that made sense 😓 

Also here's a gif of updated functionality (cause git checkout too much work sometimes)
![Feb-06-2023 06-00-56](https://user-images.githubusercontent.com/47403443/216967303-7aaf2ad8-e84d-45a5-b92b-6b3d84e6b911.gif)

ZzRanger committed:
stash changes

ZzRanger committed:
improve styles for degree plan editor
